169532131 add script check circ deps

giskarda opened the pull request.

LucaPrete commented on utils/README.md:
```diff
@@ -80,3 +80,20 @@ To synchronize users run:
+1. Create a python virtual environment
+2. Enable the environment
+2. Install required libries
+3. Run the script
```
Please, use * for lists instead to be consistent with the  format of the rest of the READMEs

LucaPrete commented on utils/README.md:
```diff
@@ -80,3 +80,20 @@ To synchronize users run:
+
+## Check for circular dependencies in live modules
+
+This is useful if we are building an environment from scratch after a long while.
```
This will be used primarily in the CI process. Can we put instead what the script does?

LucaPrete commented on utils/README.md:
```diff
@@ -80,3 +80,20 @@ To synchronize users run:
+2. Install required libries
+3. Run the script
+
+### Example:
```
Can you please substitute this with a unique block as the one below?
```
...
```

LucaPrete commented on utils/check_live_circular_dependencies.py:
```diff
@@ -0,0 +1,33 @@
+parser.add_argument('-o', '--output', help='output file')
+
+args = parser.parse_args()
+
```
please remove one of the two empty lines

ioggstream replied:
@LucaPrete @giskarda can't we delegate all that syntax stuff  to `black` ?

giskarda replied:
sure thing.

LucaPrete commented on utils/check_live_circular_dependencies.py:
```diff
@@ -0,0 +1,33 @@
+                        dot.edge(os.path.join(root, name), dep)
+                except KeyError:
+                    dot.node(name, name)
+
```
please remove one of the two empty lines

giskarda commented:
I'm going to answer in line here (I dont think I can comment your comment)

> this change doesn't come from the latest master! Please, use the updated master as starting point of your change

Ok!

> the script parses also the .terragrunt-cache folder. This must be avoided.

Fixed.

> The script produces a graph that is barely readable - either we fix the graph or the feature is useless

What you dont like? Suggestions? 

> vnet_common should depend by resource group, but I don't see any arrow

../dev/westeurope/vnet_common/terraform.tfvars has a straight line to resource_group. I can confirm

> The most important feature of the script should be find if any circular dependency is found, returning some error and exiting with -1. This will be specifically useful for CI pipelines, executed when new TF files are written. IMHO this is the primary feature.i

I've added an as_text() function that returns 1 when an circular dependency or an issue is found.

>The graph instead is a very nice to have feature, that should become optional. 

Ok 

> you should also add the following exclusions in the gitignore:
utils/live
utils/live.pdf
utils/venv/

This depends on what the user wants as a name and how it likes.

LucaPrete commented:
Answers **inline**

> this change doesn't come from the latest master! Please, use the updated master as starting point of your change

Ok!

> the script parses also the .terragrunt-cache folder. This must be avoided.

Fixed.

> The script produces a graph that is barely readable - either we fix the graph or the feature is useless

What you dont like? Suggestions? 
**I'll send you a screnshot of what I see on my screen to let you better understand what I mean. The graph is barely visible and you have to scroll horizzontally back and forth to see associations between the vertex of the graph. Maybe removing .terragrunt-cached folders made things a little bit better**

> vnet_common should depend by resource group, but I don't see any arrow

../dev/westeurope/vnet_common/terraform.tfvars has a straight line to resource_group. I can confirm

**ok**

> The most important feature of the script should be find if any circular dependency is found, returning some error and exiting with -1. This will be specifically useful for CI pipelines, executed when new TF files are written. IMHO this is the primary feature.

I've added an as_text() function that returns 1 when an circular dependency or an issue is found.

**Does it also say what the circular dependencies are?**

>The graph instead is a very nice to have feature, that should become optional. 

Ok 

> you should also add the following exclusions in the gitignore:
utils/live
utils/live.pdf
utils/venv/

This depends on what the user wants as a name and how it likes.

**venv would be a constant I guess. let's put instead of the live files in a chart (any better idea?). Then, let's add that and venv to .gitignore **
